dlt/common/libs/pyarrow.py

```diff
@@ -669,11 +688,24 @@ def add_constant_column(
         nullable: Whether the new column is nullable
         value: The value to fill the new column with
         index: The index at which to insert the new column. Defaults to -1 (append)
+    Note:
+        This function creates a dictionary field for the new column, which is memory-efficient
+        when the column contains a single repeated value.
+        The column is created as a DictionaryArray with int8 indices.
     """
-    field = pyarrow.field(name, pyarrow.dictionary(pyarrow.int8(), data_type), nullable=nullable)
+    dictionary = pyarrow.array([value], type=data_type)
+    zero_buffer = pyarrow.py_buffer(b"\x00" * item.num_rows)
```

Comment:
this is allocating bytes on Python heap and this array can be garbage collected. There's a high chance that arrow will create a duplicate buffer out of that. So we still are not 0 copy and our exercise here is to reduce memory usage.

Arrow does not have a python interface to operate on mem buffers. Because they are C and those are just part of this language. You can try this:
```
import pyarrow as pa, ctypes
buf = pa.allocate_buffer(size, resizable=False)
ctypes.memset(buf.address, 0, size)   # one pass, no extra allocation
```
unfortunatelly `allocate_buffer` does not guarantee zeroed memory so we need to memset. we can't get better than this